text/0000-undo-universal-impl-trait.md

```diff
@@ -0,0 +1,238 @@
+bindings, `const` and `static` declarations and is not of our matter here – because they also act as
+*existentials*, which is great. However, the former, [RFC 1951], introduced `impl Trait` in argument
+position. There were a lot of debates about the semantics that the syntax should have but
+eventually, the consensus came to a decision. The following syntax:
+
+    fn foo(x: impl Debug);
+
+Is akin to this:
+
+    fn foo<T: Debug>(x: T)
+    fn foo<T>(x: T) where T: Debug
+
+And so the feature was stabilized.
+
+This RFC wants to undo this.
+
+## Why is `impl Trait` in argument position wrong?
+
+People out there didn’t really realize what was going on. Complains and confusions on IRC, reddit
+and the Rust forum started to appear, after the stabilization, by different kind of people –
+confusion among newcomers, advanced proficient Rustaceans, etc.. Here are some – non-exhaustive –
+links:
+
+  - https://internals.rust-lang.org/t/how-is-impl-display-different-from-foo-t-display-x-t-t/7566/3
+  - https://www.reddit.com/r/rust/comments/8jfn7z/what_is_the_advantage_of_impl_trait_in_argument/
+  - https://www.reddit.com/r/rust/comments/8io4q6/impl_trait_is_here_now_what/dyu5dlm/
+  - https://github.com/rust-lang/rust/issues/44721#issuecomment-330935175
+
+[RFC 1951] is organized with *arguments* that we’re basing the current document against.
+
+### Argument from learnability
+
+> Argument [here](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md#argument-from-learnability)
+
+This argument explains that people who have an understanding of generics would try to use them in
+return position for anonymized types (existentials). This argument then stated it’s confusing
+because the only way to do this with current Rust is to use a complete different mechanism – being
+`impl Trait`. By contrast, if they have first learned `impl Trait` in argument position, trying to
+use it in return position *will just work as intended*.
+
+This argument is twisted along the lines of subjective opinion about learnability. Rust is a
+language with a fabulous set of learning resources (the first and most important one being
+[The Book]). It seems sound to state that **if** people are introduced first to `impl Trait` in
+argument position, it’s likely that they’ll understand the use in return location. However, what if
+they don’t want to express it? Existentials in return locations are already a very specific use case
+of a type system. Having a complete different mechanism is actually a feature, because we are
+talking about something that has no alternative elsewhere in Rust – i.e. pick the type in the
+callee. The initial assumption that having a different mechanism is confusing is wrong because
+`impl Trait` carries its own semantics and because anonymization in return type is not a trivial
+feature people will use without actually wanting it.
+
+Also, people are used to generics. People coming from C++, C#, Java etc. actually know about
```

Comment:
You've followed

> This argument is twisted along the lines of subjective opinion about learnability

With an equally subjective argument.

Reply:
People come to Rust from languages other than C++, C#, and Java. Not to mention that the entire point of `impl Trait` in argument position is that it works much closer to Java/C#. In java, you don't write void `<T extends SomeInterface> method(T arg)`, you write `void method(SomeInterface arg)`.

Reply:
Interfaces in Java are not akin to universal quantification, because those are dynamic dispatched variables. The equivalent in Rust would be `&Trait` or `dyn Trait`.

Reply:
Yup, but the entire point is that you generally don't care about the tradeoffs between a monomorphised function and a dynamically dispatched one. Similarly, you generally don't need to care about the differences between returning `Box<Foo>` and `impl Foo`. This is a "do what I mean" feature. Similar to lifetime elision, it hides details that you usually don't care about. If you end up in a situation where you do need to care about it (`T: Add<Output = T>` is a good example), or want to care about it (you're in a situation where binary size is very important), then you learn the underlying mechanisms.

Reply:
> You generally

Are you sure about that? :)

Reply:
@sgrif 
Wouldn't the actual equivalent of 
```java
void foo(SomeInterface arg)
``` 
be 
```rust
fn foo(arg: Rc<RefCell<dyn SomeTrait>>)
```
?

Because the instance is GC'ed and mutable.

So when translating Java code (or thinking) to Rust, one wouldn't just be able to write `fn foo(arg: impl SomeTrait)` instead of the java method anyway, because that would be different semantics (it takes ownership and can't mutate). One would have to re-architect the code in the translation process to use references/lifetimes and to please the borrow checker anyway. 
So `impl Trait` for args isn't really the equivalent (or what one wants) in most cases.